### DISCUSSION TITLE
Output the ordering of the stack for multiple time points to track reshuffling process

### DISCUSSION THREAD

**Discussion by Rinus-Verdonschot-UB** [asreview/asreview#1801] [Ideas]
Hello all,

To assess/confirm the (obviously great) value of ASReview we would like to compare a diverse range of ASReview simulations (e.g. simulations with golden hits, other simulations with X relevant and X non-relevant priors, and another simulation with 1% of the total dataset as Boetje/vd Schoot recommended in the SAFE procedure).

If possible, we would like to see the ordering of the stack of articles at time point 1 (the first iteration of the model), another time point X (to be determined by the user) and the time point that the model stops (last iteration). So we actually would like to be able to track the change in the ordering of the stack over time. Is this possible? And if so, how? It now seems that we can only see the output of the last iteration of the model.

I understand that outputting the whole re-schuffling of the stack for each time point is infeasible as the datafile would become too big. However, it would be nice to get the first time point (right after the seeding), the option of observing the ordering at other time points we could specify, and the last iteration (which we do already get now). Then we could have more insight how the shuffling process takes places over time and what exactly is included at important time points.

Kind regards,

Rinus Verdonschot
University Library - Maastricht University

**Comment by jteijema** [comment]
Hi! That sounds like an interesting idea! I was just answering #1802 about this same thing. We've been calling it the rank order stability (a measure of similarity between each iteration of the stack), but no definitive answers have been found yet.

Personally I don't think outputting the stack every iteration will be too big of a datafile if just for this research. 2000 records, times 2000 iterations, gives 4M lines, that's very doable on a modern device.

Keep us updated, this sounds very interesting.

As to how I'd achieve it; data scientists aren't known for their elegance but for getting things done quick. I'd just duplicate one of the classifiers and inject some code that outputs the rank order as soon as it's generated. Store it somewhere every iteration and then calculate the similarity.

**Reply by PeterLombaers** [reply]
Indeed, this isn't supported right now, though you should be able to hack something together with your approach @jteijema. But if you want the ranking and not the relevance scores, you probably want to make an extension of the query strategy, not the classifier. See [readthedocs](https://asreview.readthedocs.io/en/latest/extensions_dev.html) for more info about making extensions. Something like this should work (I didn't test it though!):

```
from asreview.models.query import MaxQuery
import json

class MaxQueryWithSave(MaxQuery):
    name = "max_with_save"

    def query(self, *args, **kwargs):
        ranked_records = super().query(*args, **kwargs)

        with open("/hardcoded/filepath/ranking.json", "a") as f:
            json.dump(ranked_records.tolist(), f)
            f.write("\n")
        
        return ranked_records
```

**Reply by jteijema** [reply]
@PeterLombaers Nice! Yes, you're totally right. Query strategy will be the place to see the ranking.
